### ISSUE TITLE
[Bug]: quota S3 storage wrong WITH Q&D FIX!

### ISSUE THREAD

**Issue by mrAceT** [nextcloud/server#34422]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Primary storage S3

It worked fine, but after a while (again....) new users get a fixed quota of 2 GB and no matter what I try I can't get that changed

### Steps to reproduce

1. create user
2. change quota
3. refresh page....

### Expected behavior

a 100% working quota system for S3 ;)

### Installation method

Community Web installer on a VPS or web space

### Operating system

RHEL/CentOS

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

MySQL

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - apporder: 0.15.0
  - calendar: 3.5.0
  - circles: 24.0.1
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contacts: 4.2.2
  - contactsinteraction: 1.5.0
  - dashboard: 7.4.0
  - dav: 1.22.0
  - deck: 1.7.1
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_external: 1.16.1
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - firstrunwizard: 2.13.0
  - groupquota: 0.1.8
  - keeweb: 0.6.9
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - mail: 1.13.8
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.1
  - oauth2: 1.12.0
  - password_policy: 1.14.0
  - photos: 1.6.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - quota_warning: 1.14.0
  - recommendations: 1.3.0
  - richdocuments: 6.2.0
  - richdocumentscode: 22.5.502
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.5
  - support: 1.7.0
  - survey_client: 1.12.0
  - systemtags: 1.14.0
  - text: 3.5.1
  - theming: 1.15.0
  - theming_customcss: 1.12.0
  - twofactor_backupcodes: 1.13.0
  - updatenotification: 1.14.0
  - user_status: 1.4.0
  - user_usage_report: 1.8.0
  - viewer: 1.8.0
  - weather_status: 1.4.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit
  - encryption
  - user_ldap
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Additional info

I have had this problem once before.

In utter dismay I (re)build my migration code to migrate back from S3 to local storage. Then the quota problem was gone..

After massive testing of my (re)built migration tool for migrating from Local > S3 I then decided to return to S3 storage.
This worked for almost exactly two months.. now I have again the exact same problem !?

Because the problem was resolved when migrating from S3 > local storage I expect the problem to reside in the filesystem handling of S3.

The strange thing is:
- when I look at the MySql entry for that user in `oc_preferences` I see the quota I set (10 Gb)!
- when I do 'occ user:setting [user] files quota'  I see the quota I set (10 Gb)!

But when I see the quota in the web interface, no matter if I'm the admin or the user I get 2 GB!
When I install the software for the network drive I get the 2 Gb quota

So effectively I set it to 10GB, it it operates at 2 GB !! Even setting it to unlimited does not change the 2GB setting!?

I have looked at the code.. to keep it a s clean as possible I started debugging via the quota info of the logged in user.. (I set it to 10 GB)

I find: apps/files/ajax/getstoragestats?dir=%2F
then: apps/files/lib/Controller/AjaxController.php: public function getStorageStats
then: apps/files/lib/Helper.php: public static function buildFileStorageStatistics
then: apps/files/lib/Controller/ViewController.php: protected function getStorageInfo
then: lib/private/legacy/OC_Helper.php: public static function getStorageInfo

There I get stuc in this part:
`		$quota = \OCP\Files\FileInfo::SPACE_UNLIMITED;
		$mount = $rootInfo->getMountPoint();
		$storage = $mount->getStorage();
		$sourceStorage = $storage;`

I can't find the "getQuota" where it goes wrong.. here:
`		// TODO: need a better way to get total space from storage
		if ($sourceStorage->instanceOfStorage('\OC\Files\Storage\Wrapper\Quota')) {
			/** @var \OC\Files\Storage\Wrapper\Quota $storage */
			$quota = $sourceStorage->getQuota();
$logger->warning("getStorageInfo testET3: includeExtStorage:".$includeExtStorage." quota:".$quota,[]);
		}`

Here I get the wrong quota.. can someone tell me where this function is in the code?

Al that  darn "getQuota" has to do is return the quota set in `oc_preferences` where  `userid` is the user and return the `configkey` value for 'quota'.

Is there anyone who can point me in the right direction?

If needed I'll "hack the system" so that I'll at least get the right quota.. But I need to know where that darn "getQuota" is at (there are quite a few functions of that name in the source code....)

AD: the version is now the latest 24.0.5, but this problem has been here for at least half a year..

**Comment by mrAceT** [comment]
I do apologize, I see now that I opened a a ticket on the first of august about my previous trouble:
https://github.com/nextcloud/server/issues/33263

What to do, merge?

I do would like to respond to my suspicion about opcache. That does seem to delay changing the quota. But now I have tested by turning off opcache and altering the quota. no effect.

Also the problem is now with an other account, not the first account, so there does not seem to be a link with the primary/first (admin) account.

The only odd thing is that, again, the wrong quota value that gets stuc is that 2 Gb value..

**Comment by mrAceT** [comment]
Since this has been bugging me on & off for over half a year now I have spent the better part of the day figuring this out.. and I have (admitted Quick and Dirty) fixed the problem:

in lib/private/Files/Storage/Wrapper/Quota.php:

added: use OC_Util; (below the other use....)

in public function __construct:
below $this->quota     = $parameters['quota']; add:
```
if (is_object($parameters['storage']->getUser())) {
$this->quotaBKP= OC_Util::getUserQuota($parameters['storage']->getUser());
  if ($this->quotaBKP > 0 ) {
    $this->quota   = $this->quotaBKP;
    $this->quotaBKP= $parameters['quota'];
  }
}
```

Somewhere (I did not find where...) the parameter[quota] gets wrongly set to 2GB.. I now simply re-read that quota.. It's not pretty but at least my headache about the subject is gone now.

**Comment by mrAceT** [comment]
I'll re-open it, hoping that a better and more permanent fix will be implemented

**Comment by mrAceT** [comment]
Oh a small addendum;

Now that I have been digging in the code, in the earlier mentioned function:
lib/private/Files/Storage/Wrapper/Quota.php:

The problem of changed quota's not being seen for a while is because in the `public static function getStorageInfo` there is a caching call to "ICacheFactory".. when I turn that off, That problem is gone.. is there something not being updated to that cache when the quota is changed?

**Comment by CarlSchwan** [comment]
Thanks for investigating @mrAceT 

That the issue is coming from the cache is pretty weird since the cache should be cleared every 5 min. What cache engine are you using? Redis/memcached/apcu?

Something that I can see is that if we fail to calculate quota for a folder we instead get the quota for the root and save that as quota for a sub folders. This is probably an issue

**Comment by mrAceT** [comment]
> cleared every 5 min
If memory serves me well,, I do think that the wrong/old setting is indeed gone after say 5 minutes.
But that does keep the question valid, why isn't the cache updated upon altering the quota?

> What cache engine are you using?
Is 'opcache' an answer to your question?

> get the quota for the root
Strange, I purposely set all quota's that might have influence to anything but 2Gb.. so there wasn't any user in the system with a 2Gb quota.. but still that quota was 2Gb!

**Comment by mrAceT** [comment]
AD: last night I upgraded my Nextcloud to version 24.0.6

As expected the "quota problem" described above returned. I implemented the fiex I described above (https://github.com/nextcloud/server/issues/34422#issuecomment-1267078757) and after a few minutes the correct quotas were back again.

So my "Q&D fix" for this quota problem also works in version 24.0.6

**Comment by phoenixwarrior95** [comment]
I've tried to apply the Q&D fix. Running 25.0.2 NC Hub 3.

The construct function is as follows:

`        public function __construct($parameters) {
                parent::__construct($parameters);
                $this->quota = $parameters['quota'] ?? null;
                $this->quotaCallback = $parameters['quotaCallback'] ?? null;
                $this->sizeRoot = $parameters['root'] ?? '';
                $this->config = \OC::$server->get(SystemConfig::class);
        }`
        
        I added in the block of code
        
        **`        public function __construct($parameters) {
                parent::__construct($parameters);
                $this->quota = $parameters['quota'] ?? null;
                if (is_object($parameters['storage']->getUser())) {
                $this->quotaBKP= OC_Util::getUserQuota($parameters['storage']->getUser());
                    if ($this->quotaBKP > 0 ) {
                    $this->quota   = $this->quotaBKP;
                    $this->quotaBKP= $parameters['quota'];
                    }
                 }
                $this->quotaCallback = $parameters['quotaCallback'] ?? null;
                $this->sizeRoot = $parameters['root'] ?? '';
                $this->config = \OC::$server->get(SystemConfig::class);
        }`**
        
Sorry for my lack of coding experience, but I do get an internal 500, meaning that I either messed up somewhere in the code or something else happened. The original problem I'm facing is similar to yours, quota on the bottom left (next to Deleted Files) is not updating right away. I have set cron to update every minute instead of 5, and for some reason it doesnt update. However, I found out that everytime I go to the User Settings after doing `sudo -u www-data php /var/www/nextcloud/occ groupquota:set Default 50GB` and placing different values for the share quota, it updates instantly. So there might be something in the user settings page that updates everytime you go to that page.

In the meantime, can anyone tell me what Im doing wrong in Quota.php? Thanks.

**Comment by mrAceT** [comment]
I've done the update to 25.0.2 and altered the "Q&D fix" a little bit, this should help you enough:
![image](https://user-images.githubusercontent.com/29061116/207434322-d257dfbc-4132-4ecb-b8c3-c73de03b83c1.png)

```
 		$this->quotaBKP= null;
		if (is_object($this->getWrapperStorage()->getUser())) {
			$this->quotaBKP= OC_Util::getUserQuota($this->getWrapperStorage()->getUser());
			if ($this->quotaBKP > 0 ) {
				$this->quota   = $this->quotaBKP;
				$this->quotaBKP= $parameters['quota'] ?? null;
			}
 		}
```

**Comment by mrAceT** [comment]
did you reset opcache? (php: opcache_reset();)
did you restart apache/nginx?
did you restart php?

(Buffering is a b*tch..)

Otherwise, restore your original, error gone? I didn't change anything else then that after the upgrade.. can you give me the error?

**Comment by mrAceT** [comment]
Did you add `use OC_Util; ` ?

**Comment by phoenixwarrior95** [comment]
Yes, I did use OC_Util;

**Comment by mrAceT** [comment]
Please check https://github.com/nextcloud/server/issues/34422#issuecomment-1349628070 will update to 25.0.3.. but 25.0.2 still has the bug, so I'm not holding my hopes up..

**Comment by mrAceT** [comment]
I am now at version 25.0.9 (haven't upgraded to 26 yet)

And I ran into the same problem again. I have (again) performed "my hack" and the problem was fixed.

"that hack/Q&D fix" : https://github.com/nextcloud/server/issues/34422#issuecomment-1267078757

**Comment by mrAceT** [comment]
I am now at version 27.1.6 (haven't upgraded to 27.1.7 yet)

Same problem again.. same solution again..

**Comment by joshtrichards** [comment]
Hi @mrAceT - I'm unable to reproduce this. I also don't see evidence that there have been similar reports.

Are you still experiencing this?

If so:

- What is the output of `occ user:info <user>`?
- What was the exact (faulty) value reported by your debug logging code?
- Are you, by chance, on a 32-bit system? 
- Are you able to reproduce this in a clean installation (e.g. a test install)

**Comment by nextcloud-command** [comment]
This issue has been automatically marked as stale because it has not had recent activity and seems to be missing some essential information. It will be closed if no further activity occurs. Thank you for your contributions.

**Comment by zmberber** [comment]
I also have this problem.  Is there nothing one can do to fix this?

**Comment by mrAceT** [comment]
@zmberber I know it's not pretty, but it does get the job done, have you seen my "Q&D Fix" : https://github.com/nextcloud/server/issues/34422#issuecomment-1267078757

Also, I think it will help if you would if you told which versions (also of PHP) you are running.

**Comment by zmberber** [comment]
@mrAceT thank you for replying!  it turned out to actually be a problem with a plugin `oidc_login`, where if you set a default quota, you apparently cannot change the quota within the webinterface, but only through the oidc user attribute `quota`.  i had removed in `config.php` whatever enabled this behaviour but it didn't stop, i was able to manually disable it by setting `oidc_login_default_quota` to `0` in `config.php`, which actively disables the feature.  then it worked again.  i wonder where this value was saved in the database, because if it is not set in config.php, it must be set as a value somewhere else.

- php version 8.3.20
- nextcloud 31.0.4

**Comment by mrAceT** [comment]
The strangest thing.. this problem oddly pops back up every now and then.. The last time I implemented this fix was with version 30.0.4 and recently I got complaints again from a user.. I have again implemented this Quick & Dirty fix (https://github.com/nextcloud/server/issues/34422#issuecomment-1267078757) on **32.0.5**.

And again this fixed the quota problem !

So somehow, somewhere this bug seems to be gone for a time and then pops up again.. and the only way I know to fix it is to implement my bugfix again!

To my knowledge the only "little odd thing" I have in my setup is that I have **S3** as the primary storage.. but that isn't that odd.. is it?

PS: this NextCloud is now running on PHP 8.2 I will upgrade to 8.3 in a while (but this bug started out on 7.4.. so I am not thinking the primary problem lies there..)

PS2: @zmberber sorry for the late response (your problem was fixed so I didn't pay much attention to the rest of your message)
Interresting.. but I don't use 'oidc' so that isn't the cause at my end.. that is still a bit of a mystery I guess..
